plugins/volume/org.xen.xapi.storage.zfs-ng/volume.py

```diff
@@ -0,0 +1,252 @@
+                    if vdi.volume.parent_id == None:
+                        snap_volume = db.insert_new_volume(vdi.volume.vsize, vdi.image_type)
+                    else:
+                        snap_volume = db.insert_child_volume(vdi.volume.parent_id,
+                                                         vdi.volume.vsize)
+                    db.set_volume_as_snapshot(snap_volume.id)
+                    db.insert_vdi(vdi.name, vdi.description,
+                                  snap_uuid, snap_volume.id, False)
+
+                    db.update_volume_parent(vdi.volume.id, snap_volume.id)
+
+                    result_volume_id = str(snap_volume.id)
+                    ZFSUtil.snapshot(dbg, str(snap_volume.id),
+                                 os.path.basename(sr) + '/'+ str(vdi.volume.id), False)
+                    path = os.path.basename(sr) + '/'+ str(vdi.volume.id) + '@' + str(snap_volume.id)
+                    ZFSUtil.clone(dbg, path, os.path.basename(sr) + '/' + str(snap_volume.id))
+                    ZFSUtil.promote(dbg, os.path.basename(sr) + '/' + str(snap_volume.id))
+        psize = 0
+        snap_uri = cb.getVolumeUriPrefix(opq) + snap_uuid
+        return {
+            'uuid': snap_uuid,
+            'key': snap_uuid,
+            'name': result_volume_id,
+            'description': vdi.description,
+            'read_write': False,
+            'virtual_size': vdi.volume.vsize,
+            'physical_utilisation': psize,
+            'uri': [image_format.uri_prefix + snap_uri],
+            'keys': {},
+            'sharable': False
+        }
+
+    # clone only works on snapshots
+    # fails otherwise
+    def clone(self, dbg, sr, key):
+        snap_uuid = str(uuid.uuid4())
+        cb = self.callbacks
+        with VolumeContext(cb, sr, 'w') as opq:
+            result_volume_id = ''
+            with PollLock(opq, 'gl', cb, 0.5):
+                with cb.db_context(opq) as db:
+                    vdi = db.get_vdi_by_id(key)
+                    if vdi.volume.snap == 0:
+                        log.error('Only snapshots can be cloned!')
+                        raise
+                    image_format = ImageFormat.get_format(vdi.image_type)
+                    image_utils = image_format.image_utils
+                    cloned_volume = db.insert_new_volume(vdi.volume.vsize, vdi.image_type)
+
+                    # clone parent is the snapshot
+                    db.insert_vdi(
+                        vdi.name, vdi.description, snap_uuid, cloned_volume.id, False)
+                    db.update_volume_parent(cloned_volume.id, vdi.volume.id)
+
+                    result_volume_id = str(cloned_volume.id)
+                    snap_path = os.path.basename(sr) + '/'+ str(vdi.volume.id) + '@' + str(vdi.volume.id)
+                    clone_path = os.path.basename(sr) + '/'+ result_volume_id
+                    ZFSUtil.clone(dbg, snap_path, clone_path)
+
+        psize = 0
```

Comment:
Always 0?

Reply:
I fixed this by using the value from get_vsize. I do not remember if clone volumes consumes space.